File: java/patterns/src/main/java/org/prototypepatterns/PatternRegistry.java
package org.prototypepatterns;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import org.prototypepatterns.behavioral.chainofresponsibility.ChainOfResponsibilityPattern;
import org.prototypepatterns.behavioral.command.CommandPattern;
import org.prototypepatterns.behavioral.iterator.IteratorPattern;
import org.prototypepatterns.behavioral.mediator.MediatorPattern;
import org.prototypepatterns.behavioral.memento.MementoPattern;
import org.prototypepatterns.behavioral.observer.ObserverPattern;
import org.prototypepatterns.behavioral.state.StatePattern;
import org.prototypepatterns.behavioral.strategy.StrategyPattern;
import org.prototypepatterns.behavioral.templatemethod.TemplateMethodPattern;
import org.prototypepatterns.creational.abstractfactory.AbstractFactoryPattern;
import org.prototypepatterns.creational.builder.BuilderPattern;
import org.prototypepatterns.creational.factorymethod.FactoryMethodPattern;
import org.prototypepatterns.creational.lazyinitialization.LazyInitializationPattern;
import org.prototypepatterns.creational.objectpool.ObjectPoolPattern;
import org.prototypepatterns.creational.prototypepattern.PrototypePattern;
import org.prototypepatterns.creational.singleton.SingletonPattern;
import org.prototypepatterns.structural.adapter.AdapterPattern;
import org.prototypepatterns.structural.bridge.BridgePattern;
import org.prototypepatterns.structural.composite.CompositePattern;
import org.prototypepatterns.structural.compositeentity.CompositeEntityPattern;
import org.prototypepatterns.structural.decorator.DecoratorPattern;
import org.prototypepatterns.structural.facade.FacadePattern;
import org.prototypepatterns.structural.flyweight.FlyweightPattern;
import org.prototypepatterns.structural.proxy.ProxyPattern;

public final class PatternRegistry {
    public record PatternEntry(String id, String category, Supplier<Object> demo) {
    }

    private static final List<PatternEntry> PATTERNS = List.of(
            new PatternEntry("singleton", "creational", () -> SingletonPattern.demo()),
            new PatternEntry("factory-method", "creational", () -> FactoryMethodPattern.demo()),
            new PatternEntry("abstract-factory", "creational", () -> AbstractFactoryPattern.demo()),
            new PatternEntry("builder", "creational", () -> BuilderPattern.demo()),
            new PatternEntry("prototype", "creational", () -> PrototypePattern.demo()),
            new PatternEntry("object-pool", "creational", () -> ObjectPoolPattern.demo()),
            new PatternEntry("lazy-initialization", "creational", () -> LazyInitializationPattern.demo()),
            new PatternEntry("adapter", "structural", () -> AdapterPattern.demo()),
            new PatternEntry("decorator", "structural", () -> DecoratorPattern.demo()),
            new PatternEntry("composite", "structural", () -> CompositePattern.demo()),
            new PatternEntry("facade", "structural", () -> FacadePattern.demo()),
            new PatternEntry("proxy", "structural", () -> ProxyPattern.demo()),
            new PatternEntry("bridge", "structural", () -> BridgePattern.demo()),
            new PatternEntry("flyweight", "structural", () -> FlyweightPattern.demo()),
            new PatternEntry("composite-entity", "structural", () -> CompositeEntityPattern.demo()),
            new PatternEntry("observer", "behavioral", () -> ObserverPattern.demo()),
            new PatternEntry("strategy", "behavioral", () -> StrategyPattern.demo()),
            new PatternEntry("command", "behavioral", () -> CommandPattern.demo()),
            new PatternEntry("state", "behavioral", () -> StatePattern.demo()),
            new PatternEntry("template-method", "behavioral", () -> TemplateMethodPattern.demo()),
            new PatternEntry("chain-of-responsibility", "behavioral", () -> ChainOfResponsibilityPattern.demo()),
            new PatternEntry("iterator", "behavioral", () -> IteratorPattern.demo()),
            new PatternEntry("mediator", "behavioral", () -> MediatorPattern.demo()),
            new PatternEntry("memento", "behavioral", () -> MementoPattern.demo()));

    private static final Map<String, PatternEntry> PATTERN_BY_ID = PATTERNS.stream()
            .collect(Collectors.toMap(PatternEntry::id, entry -> entry));

    private PatternRegistry() {
    }

    public static String normalizePatternId(String patternId) {
        return patternId.strip().toLowerCase().replace('_', ' ').replaceAll("\\s+", "-");
    }

    public static List<PatternEntry> list() {
        return PATTERNS;
    }

    public static Object runPattern(String patternId) {
        String normalized = normalizePatternId(patternId);
        PatternEntry entry = PATTERN_BY_ID.get(normalized);
        if (entry == null) {
            throw new IllegalArgumentException("Unknown pattern: " + patternId);
        }
        return entry.demo().get();
    }

    public static Map<String, Object> runAllPatterns() {
        Map<String, Object> results = new LinkedHashMap<>();
        for (PatternEntry entry : PATTERNS) {
            results.put(entry.id(), entry.demo().get());
        }
        return results;
    }
}

File: csharp/patterns/src/PrototypePatterns/PatternRegistry.cs
using System.Text.RegularExpressions;
using PrototypePatterns.Behavioral.ChainOfResponsibility;
using PrototypePatterns.Behavioral.Command;
using PrototypePatterns.Behavioral.Iterator;
using PrototypePatterns.Behavioral.Mediator;
using PrototypePatterns.Behavioral.Memento;
using PrototypePatterns.Behavioral.Observer;
using PrototypePatterns.Behavioral.State;
using PrototypePatterns.Behavioral.Strategy;
using PrototypePatterns.Behavioral.TemplateMethod;
using PrototypePatterns.Creational.AbstractFactory;
using PrototypePatterns.Creational.Builder;
using PrototypePatterns.Creational.FactoryMethod;
using PrototypePatterns.Creational.LazyInitialization;
using PrototypePatterns.Creational.ObjectPool;
using PrototypePatterns.Creational.Prototype;
using PrototypePatterns.Creational.Singleton;
using PrototypePatterns.Structural.Adapter;
using PrototypePatterns.Structural.Bridge;
using PrototypePatterns.Structural.Composite;
using PrototypePatterns.Structural.CompositeEntity;
using PrototypePatterns.Structural.Decorator;
using PrototypePatterns.Structural.Facade;
using PrototypePatterns.Structural.Flyweight;
using PrototypePatterns.Structural.Proxy;

namespace PrototypePatterns;

public sealed record PatternEntry(string Id, string Category, Func<object?> Demo);

public static class PatternRegistry
{
    private static readonly IReadOnlyList<PatternEntry> Patterns =
    [
        new("singleton", "creational", () => SingletonPattern.Demo()),
        new("factory-method", "creational", () => FactoryMethodPattern.Demo()),
        new("abstract-factory", "creational", () => AbstractFactoryPattern.Demo()),
        new("builder", "creational", () => BuilderPattern.Demo()),
        new("prototype", "creational", () => PrototypePattern.Demo()),
        new("object-pool", "creational", () => ObjectPoolPattern.Demo()),
        new("lazy-initialization", "creational", () => LazyInitializationPattern.Demo()),
        new("adapter", "structural", () => AdapterPattern.Demo()),
        new("decorator", "structural", () => DecoratorPattern.Demo()),
        new("composite", "structural", () => CompositePattern.Demo()),
        new("facade", "structural", () => FacadePattern.Demo()),
        new("proxy", "structural", () => ProxyPattern.Demo()),
        new("bridge", "structural", () => BridgePattern.Demo()),
        new("flyweight", "structural", () => FlyweightPattern.Demo()),
        new("composite-entity", "structural", () => CompositeEntityPattern.Demo()),
        new("observer", "behavioral", () => ObserverPattern.Demo()),
        new("strategy", "behavioral", () => StrategyPattern.Demo()),
        new("command", "behavioral", () => CommandPattern.Demo()),
        new("state", "behavioral", () => StatePattern.Demo()),
        new("template-method", "behavioral", () => TemplateMethodPattern.Demo()),
        new("chain-of-responsibility", "behavioral", () => ChainOfResponsibilityPattern.Demo()),
        new("iterator", "behavioral", () => IteratorPattern.Demo()),
        new("mediator", "behavioral", () => MediatorPattern.Demo()),
        new("memento", "behavioral", () => MementoPattern.Demo())
    ];

    private static readonly IReadOnlyDictionary<string, PatternEntry> PatternById =
        Patterns.ToDictionary(pattern => pattern.Id, pattern => pattern);

    public static string NormalizePatternId(string patternId)
    {
        var normalized = patternId.Trim().ToLowerInvariant().Replace('_', ' ');
        return Regex.Replace(normalized, @"\s+", "-");
    }

    public static IReadOnlyList<PatternEntry> List() => Patterns;

    public static object? RunPattern(string patternId)
    {
        var normalized = NormalizePatternId(patternId);
        if (!PatternById.TryGetValue(normalized, out var pattern))
        {
            throw new ArgumentException($"Unknown pattern: {patternId}", nameof(patternId));
        }

        return pattern.Demo();
    }

    public static IReadOnlyDictionary<string, object?> RunAllPatterns()
    {
        var results = new Dictionary<string, object?>();
        foreach (var pattern in Patterns)
        {
            results[pattern.Id] = pattern.Demo();
        }

        return results;
    }
}
